drivers/flash/flash_stm32.c

```diff
@@ -65,11 +64,7 @@ LOG_MODULE_REGISTER(flash_stm32, CONFIG_FLASH_LOG_LEVEL);
  */
 static inline void _flash_stm32_sem_take(struct device *dev)
 {
-
-#ifdef CONFIG_SOC_SERIES_STM32WBX
-	while (LL_HSEM_1StepLock(HSEM, CFG_HW_FLASH_SEMID)) {
-	}
-#endif /* CONFIG_SOC_SERIES_STM32WBX */
+	z_stm32_hsem_lock(CFG_HW_FLASH_SEMID);
 
 	k_sem_take(&FLASH_STM32_PRIV(dev)->sem, K_FOREVER);
```

Comment:
This seems wrong for a couple reasons.

First, the order is wrong.  `k_sem_take()` can put the calling thread to sleep for as long as it takes a previous invocation to complete.  Presumably that invocation has already taken the HSEM lock, so again we have a double-lock / premature unlock situation (the lock taken here will have been released by the unlock that releases the semaphore this thread is waiting for).

So at a minimum HSEM must be locked and unlocked while the thread holds the lock on the device.

Second, this means HSEM is kept locked for as long as a flash operation is in progress, which is a very long time.  It seems likely other things might invoke operations that also take and release the HSEM lock.

I suspect this driver needs to use much finer-grained locking of HSEM, just as most of the other drivers do.

I'm also concerned now that if any of the API that takes and release HSEM can be invoked from an interrupt (e.g. `gpio_pin_set()`) that the locking regiment will be broken because the nested lock operation has no effect, but the unlock will release both locks.  Not only here, but possibly everywhere else that this new API is used to lock and unlock while interrupts are enabled.

I think this PR needs an architecture overview from an STM maintainer.  It may require a different solution.

Reply:
First: Reverse order -> done.
Second: Please note that there are several distinct HSEM (cuurently 1 for Flash, 1for GPIO, 1 for EXTI, ...). So we have some granularity.
And if 

> other things might invoke operations that also take and release the HSEM lock

it must be on the same Flash HSEM. So it is no so likekely. Nevertheless, it is still possible to have both cores trying to write flash, so, specially for FLASH use case which may hold the lock for long time,  I allow to wait forever.

Concerning interrupt use case, I agree on this issue, but it is not specific to HSEM or dualcore. 
As you know there are lot of discussions around multithreading and/or interruptions (#18970 , #6184, #24511,  ...).
My view for HSEM is that: first in general we should tend to have a single driver able to access register of a specific IP (example only clock control should access to RCC registers, ...), then if this driver is protected against multithread and interruptions, then HSEM will implicitly be protect too.
But to be honest I didn't clearly understand how drivers will be protected against Interruptions.

For the specific case of gpio: `gpio_pin_set()` doesn't make use of HSEM, because it uses specific register able to address a single bit of a register. Nevertheless I noticed that it is not the case for all functions (like toggle). I will investigate to see how far I can get ride of HSEM and use this gpio specific register (in separate PR).